File: prune_saenet/estimate_adj_matrix_for_stacked_imputed_datasets.r
if (!require(miselect)) {
    quit("no", 2)
}

library(miselect)

# command args
temp_dir <- NULL
seed <- NULL

args <- commandArgs(trailingOnly=TRUE)
for (arg in args) {
    s <- strsplit(arg, "=")[[1]]
    if (length(s) < 2) next

    if (s[1] == "--temp_dir") {
        temp_dir <- paste(s[2:length(s)], collapse="=")
    } else if (s[1] == "--rs") {
        seed <- strtoi(s[2])
    }
}

if (!is.null(seed))
    set.seed(seed)

# args
X_list <- c()
causal_order <- NULL
weights <- NULL
ad_weight_type <- NULL
prior_knowledge <- NULL

# X_list
path <- file.path(temp_dir, "X_names.csv")
X_names <- read.csv(path, sep=",", header=FALSE)
X_names <- lapply(X_names, function(x) {paste(temp_dir, "/", x, sep="")})[[1]]
X_list <- lapply(X_names, read.csv, sep=",", header=FALSE)

# causal_order
path <- file.path(temp_dir, "causal_order.csv")
causal_order <- read.csv(path, sep=',', header=FALSE)
causal_order <- causal_order + 1

# ad_weight_type (1se or min)
path <- file.path(temp_dir, "ad_weight_type.csv")
ad_weight_type <- read.csv(path, sep=',', header=FALSE)[1, 1]

# weights
path <- file.path(temp_dir, "weights.csv")
weights <- read.csv(path, sep=",", header=FALSE)
weights <- as.vector(weights)[[1]]

# prior_knowledge
path <- file.path(temp_dir, "prior_knowledge.csv")
if (file.exists(path))
    prior_knowledge <- read.csv(path, sep=',', header=FALSE)

# params
n_imputation <- length(X_list)
n_sample <- dim(X_list[[1]])[[1]]
n_feature <- dim(X_list[[1]])[[2]]

# estimated adjacency matrix
B <- data.frame(matrix(0, nrow=n_feature, ncol=n_feature))

for (i in 2:dim(causal_order)[1]) {
    predictors <- causal_order[1:i - 1, 1]
    target <- causal_order[i, 1]

    # apply prior_knowledge
    if (!is.null(prior_knowledge)) {
        drop_targets <- c()

        for (j in 1:length(predictors))
            if (prior_knowledge[target, predictors[j]] == 0)
                drop_targets <- append(-1 * j, drop_targets)

        if (length(drop_targets) > 0)
            predictors <- predictors[drop_targets]
    }
    
    if (length(predictors) == 0) next

    # make dataset
    x <- lapply(X_list, function(x_) { as.matrix(x_[, predictors]) })
    y <- lapply(X_list, function(x_) { as.vector(x_[, target]) })

    # options
    pf <- rep(1, length(predictors))

    # run to search adaptive weights
    adWeight_cv <- rep(1, length(predictors))
    
    fit <- saenet(x, y, pf, adWeight_cv, weights)
    CV <- cv.saenet(x, y, pf, adWeight_cv, weights, lambda=fit$lambda)
    coef_cv <- fit$coef[, 1, ]

    lambda_list_cv <- fit$lambda
    
    if (ad_weight_type == "min") {
        key <- "lambda.min"
    } else if (ad_weight_type == "1se") {
        key <- "lambda.1se"
    }
    lambda_cv <- CV[[key]]
    
    for (i in 1:length(lambda_list_cv)) {
        if (abs(lambda_list_cv[[i]] / lambda_cv - 1.0) < 1.0e-6)
            n_lambda_cv = i
    }
    
    # run to search beta
    beta_hat_cv <- coef_cv[n_lambda_cv, 2:dim(coef_cv)[[2]]]
    abs_beta_hat_cv <- abs(beta_hat_cv) + 1.0 / (n_sample * n_imputation)
    nu <- log(n_feature) / log(n_sample * n_imputation)
    gamma <- ceiling(2 * nu / (1 - nu) + 1)
    adWeight_cv <- abs_beta_hat_cv ** (-gamma)
    
    fit <- saenet(x, y, pf, adWeight_cv, weights)
    CV <- cv.saenet(x, y, pf, adWeight_cv, weights, lambda=fit$lambda)
    
    lambda_list_cv <- fit$lambda
    lambda_cv <- CV[[key]]
    for (i in 1:length(lambda_list_cv)) {
        if (abs(lambda_list_cv[[i]] / lambda_cv - 1.0) < 1.0e-6)
            n_lambda_cv = i
    }
    beta_hat_cv <- coef_cv[n_lambda_cv, 2:dim(coef_cv)[[2]]]
    
    B[target, predictors] <- beta_hat_cv
}

path <- file.path(temp_dir, "result_adj_mat.csv")
write.csv(B, path, row.names=F)

quit("no", 0)

File: prune_saenet/_estimate_adj_matrix_for_stacked_imputed_datasets.py
import os
import shutil
import tempfile
import subprocess

import numpy as np
from sklearn.utils import check_array, check_scalar


def estimate_adj_matrix_for_stacked_imputed_datasets(X_list, causal_order, is_discrete=None, ad_weight_type="min", weights=None, prior_knowledge=None, seed=None):
    """ estimation of pruned adjacency matrices assuming a common pattern

    Parameters
    ----------
    X_list : list, shape [X, ...]
        Multiple datasets for training, where ``X`` is an dataset.
        
        The shape of ''X'' is (n_samples, n_features),
        where ``n_samples`` is the number of samples and ``n_features`` is the number of features.
    causal_order : list, length = n_features
        The causal order of X_list.
    is_discrete : array-like of shape (n_features, ), optional (default=None)
        ``is_discrete[i]`` specifies whether the i-th feature is discrete or not.
    ad_weight_type : string, "min" or "1se", optional (default="min")
        The name of the method to decide the adaptive weights.
    weights : array-like of shape (n_samples, ), optional (default=None)
        The proportion of the missing data for each row in X. 
    prior_knowledge : array-like, shape (n_features, n_features), optional (default=None)
        The prior knowledge used for the causal discovery, where ``n_features`` is the number of features.

        The elements of prior knowledge matrix are defined as follows [1]_:

        * ``0`` : :math:`x_i` does not have a directed path to :math:`x_j`
        * ``1`` : :math:`x_i` has a directed path to :math:`x_j`
        * ``-1`` : No prior knowledge is available to know if either of the two cases above (0 or 1) is true.
    seed : int, optional (default=None)
        The seed for random numbers.
    
    Returns
    -------
    adjacency_matrix : array-like
        The shape of ``adjacency_matrix`` is (n_features, n_features), where
        ``n_features`` is the number of features.
    """
    # check arguments
    X_list = check_array(X_list, allow_nd=True)
    if len(X_list.shape) != 3:
        raise ValueError("X_list must be 3-dimensional.")
    if 0 in X_list.shape:
        raise ValueError("The shape of X_list mustn't contains 0.")
    
    n_imputations, n_samples, n_features = X_list.shape
    
    causal_order = check_array(causal_order, ensure_2d=False)
    if len(causal_order) != n_features:
        raise ValueError("The length of causal_order must equal to n_features.")
    if not np.all(np.sort(causal_order) == np.arange(n_features)):
        raise ValueError("Elements of causal_order must be unique between 0 and n_features.")

    if is_discrete is not None:
        is_discrete = check_array(is_discrete, dtype=bool, ensure_2d=False)
        if len(is_discrete) != n_features:
            raise ValueError("The length of is_discrete must be equal to n_features.")

        is_discrete = list(map(lambda s: str(s).upper(), is_discrete))
        
    ad_weight_type = check_scalar(ad_weight_type, "ad_weight_type", str)
    if not (ad_weight_type == "min" or ad_weight_type == "1se"):
        raise ValueError("ad_weight_type must be \"min\" or \"1se\".")
    
    if weights is not None:
        weights = check_array(weights, ensure_2d=False)
        if weights.shape != (n_samples, ):
            raise ValueError("The length of weights must be equal to n_samples.")
    else:
        weights = np.ones(n_samples)
    
    if prior_knowledge is not None:
        prior_knowledge = check_array(prior_knowledge)
        if prior_knowledge.shape != (n_features, n_features):
            raise ValueError("The shape of prior_knowledge must be (n_features, n_features).")
    
    if seed is not None:
        seed = check_scalar(seed, "seed", int)
    
    try:
        temp_dir = tempfile.mkdtemp()
        
        path = os.path.join(os.path.dirname(__file__), "estimate_adj_matrix_for_stacked_imputed_datasets.r")
        
        args = [f"--temp_dir={temp_dir}"]
        if seed is not None:
            args.append(f"--rs={seed}")
        
        # write data
        X_names = []
        for i, X in enumerate(X_list):
            X_name = f"X_list_{i:d}.csv"
            np.savetxt(temp_dir + "/" + X_name, X, delimiter=",")
            X_names.append([X_name])
        
        # write params
        np.savetxt(temp_dir + "/X_names.csv", X_names, delimiter=",", fmt="%s")
        np.savetxt(temp_dir + "/causal_order.csv", causal_order, delimiter=",", fmt="%d")
        if is_discrete is not None:
            np.savetxt(temp_dir + "/is_discrete.csv", [is_discrete], delimiter=",", fmt="%s")
        np.savetxt(temp_dir + "/ad_weight_type.csv", [ad_weight_type], delimiter=",", fmt="%s")
        np.savetxt(temp_dir + "/weights.csv", weights, delimiter=",")
        if prior_knowledge is not None:
            np.savetxt(temp_dir + "/prior_knowledge.csv", prior_knowledge, delimiter=",")
        
        # run
        ret = subprocess.run(["Rscript", path, *args], capture_output=True)
        if ret.returncode != 0:
            print(ret.stdout.decode())
            print(ret.stderr.decode())

        # retrieve result
        adjacency_matrix = np.loadtxt(temp_dir + "/result_adj_mat.csv", delimiter=",", skiprows=1)
    except FileNotFoundError as e:
        raise RuntimeError("Rscript is not found.")
    except BaseException as e:
        raise RuntimeError(str(e))
    finally:
        if os.path.exists(temp_dir):
            shutil.rmtree(temp_dir)
    
    return adjacency_matrix
